fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -2294,7 +2577,190 @@ List<TExecPlanFragmentParams> toThrift(Set<TUniqueId> inFlightInstanceIds,
+         *
+         * @param uniqueParams           The destination unique thrift params.
+         * @param fragmentIndex          The index of this instance in this.instanceExecParams.
+         * @param instanceExecParam      The instance param.
+         * @param enablePipelineEngine Whether enable pipeline engine.
+         */
+        private void toThriftForUniqueParams(TExecPlanFragmentParams uniqueParams, int fragmentIndex,
+                                             FInstanceExecParam instanceExecParam, boolean enablePipelineEngine)
+                throws Exception {
+            uniqueParams.setProtocol_version(InternalServiceVersion.V1);
+            uniqueParams.setBackend_num(instanceExecParam.backendNum);
+            if (enablePipelineEngine) {
+                if (instanceExecParam.isSetPipelineDop()) {
+                    uniqueParams.setPipeline_dop(instanceExecParam.pipelineDop);
+                } else {
+                    uniqueParams.setPipeline_dop(fragment.getPipelineDop());
+                }
+            }
+
+            // add instance number in file name prefix when export job
+            if (fragment.getSink() instanceof ExportSink) {
+                ExportSink exportSink = (ExportSink) fragment.getSink();
+                if (exportSink.getFileNamePrefix() != null) {
+                    exportSink.setFileNamePrefix(exportSink.getFileNamePrefix() + fragmentIndex + "_");
+                }
+            }
+
+            if (!uniqueParams.isSetFragment()) {
+                uniqueParams.setFragment(fragment.toThriftForUniqueFields());
+            }
+
+            /*
+             * For MultiCastDataFragment, output only send to local, and the instance is keep
+             * same with MultiCastDataFragment
+             * */
+            if (fragment instanceof MultiCastPlanFragment) {
+                List<List<TPlanFragmentDestination>> multiFragmentDestinations =
+                        uniqueParams.getFragment().getOutput_sink().getMulti_cast_stream_sink().getDestinations();
+                List<List<TPlanFragmentDestination>> newDestinations = Lists.newArrayList();
+                for (List<TPlanFragmentDestination> destinations : multiFragmentDestinations) {
+                    Preconditions.checkState(instanceExecParams.size() == destinations.size());
+                    TPlanFragmentDestination ndes = destinations.get(fragmentIndex);
+
+                    Preconditions.checkState(ndes.getServer().equals(toRpcHost(instanceExecParam.host)));
+                    newDestinations.add(Lists.newArrayList(ndes));
+                }
+
+                uniqueParams.getFragment().getOutput_sink().getMulti_cast_stream_sink()
+                        .setDestinations(newDestinations);
+            }
+
+            if (!uniqueParams.isSetParams()) {
+                uniqueParams.setParams(new TPlanFragmentExecParams());
+            }
+            uniqueParams.params.setQuery_id(queryId);
```

Comment:
why query id is unique?

Reply:
Oh, `query_id` is a required field, let me move it to `fillRequiredFieldsToThrift`.